### PR TITLE
[3.5] bpo-36478: Use C89 for loops in backported pickle patch

### PR DESCRIPTION
**Pull request by asottile**
https://bugs.python.org/issue36478

### CODE REVIEW

**Comment by asottile** [comment]
CC @vstinner  (original patch ef33dd6036aafbd3f06c1d56e2b1a81dae3da63c)

**Review by pganssle — APPROVED** [review]
LGTM, assuming that it was a deliberate choice to not include detection of regressions of this sort as part of the CI tests, which I assume @vstinner would know.

If that was an oversight, now might be a good time to try to remedy the situation (would also make it easier to tell if this PR actually solves the bug, too).

**Comment by asottile** [comment]
fwiw, this patch was necessary to build for deadsnakes: https://github.com/deadsnakes/python3.5/commit/fe8602d3148dfaffbd18a9a1324bdcb940ce1ae0

**Comment by vstinner** [comment]
> assuming that it was a deliberate choice to not include detection of regressions of this sort as part of the CI tests, which I assume @vstinner would know.

I'm not sure of what you means by "regression". Using GCC 8.3.1 (on Fedora 29), I get 65 compiler warnings but no warnings on _pickle.c, with "./configure -C --with-pydebug CFLAGS=-O0 && make".

https://bugs.python.org/issue36478 is just a compiler warning. I don't know why it is seen as an error. For example, GCC uses ISO C99 by default.

**Comment by vstinner** [comment]
I tried gcc -std=c89, the compiler fails early on a different file:

```
gcc -pthread -c -Wno-unused-result -Wsign-compare -g -Og -Wall -Wstrict-prototypes -std=c89   -Werror=declaration-after-statement   -I. -I./Include    -DPy_BUILD_CORE -o Objects/listobject.o Objects/listobject.c
Objects/bytesobject.c: In function '_PyBytes_Format':
Objects/bytesobject.c:782:17: error: C++ style comments are not allowed in ISO C90
                 // %r is only for 2/3 code; 3 only code should use %a
                 ^
Objects/bytesobject.c:782:17: error: (this will be reported only once per input file)
```

This comment was added 4 years ago:

```
commit 62e977f1b6d52a973304db9b0268aece99cb6c42
Author: Ethan Furman <ethan@stoneleaf.us>
Date:   Wed Mar 11 08:17:00 2015 -0700

    Close issue23467: add %r compatibility to bytes and bytearray
``` 

I'm not sure which kind of problem we are trying to solve here.

I know that on Windows, "error: 'for' loop initial declarations are only allowed in C99 mode" might be an error, but at this point, nobody tried to build the latest Python 3.5 on Windows and the compilation goes fine on Linux with default GCC flags. There issue looks to be specific to compiler flags used by https://github.com/deadsnakes

**Comment by pganssle** [comment]
> I'm not sure of what you means by "regression".

Sorry, for some reason I misread the original issue and thought that you filed it, which is why I figured it was something that was detected by a buildbot.

According to [PEP 7](https://www.python.org/dev/peps/pep-0007/#c-dialect), for Python 3.5, the code should be C89 and it seems like it's one step above a pure-style refactoring if `deadsnakes` requires it to build.

**Comment by asottile** [comment]
The compiler targetted here is pretty old (but then again, so is python3.5): gcc 4.8.x

From the man page: https://gcc.gnu.org/onlinedocs/gcc-4.8.5/gcc/C-Dialect-Options.html#C-Dialect-Options

> ‘gnu89’
>    GNU dialect of ISO C90 (including some C99 features). This is the default for C code. 

In gnu89:

```console
root@5b98b67e14ee:/# gcc --version
gcc (Ubuntu 4.8.4-2ubuntu1~14.04.4) 4.8.4
Copyright (C) 2013 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.

root@5b98b67e14ee:/# gcc t.c
t.c: In function 'main':
t.c:4:5: error: 'for' loop initial declarations are only allowed in C99 mode
     for (int i = 0; i < 5; i += 1) {
     ^
t.c:4:5: note: use option -std=c99 or -std=gnu99 to compile your code
root@5b98b67e14ee:/# cat t.c 
#include <stdio.h>

int main(void) {
    for (int i = 0; i < 5; i += 1) {
        printf("%d\n", i);
    }
}
```

**Comment by thehesiod** [comment]
needed for debian:stretch builds with default gcc

File: Misc/NEWS.d/next/Build/2019-03-29-14-29-06.bpo-36478.hzyneF.rst [hunk 1]
```diff
@@ -0,0 +1 @@
+Use C89 for loops in backported pickle patch.  Patch by Anthony Sottile.
```
Comment (vstinner) [inline Misc/NEWS.d/next/Build/2019-03-29-14-29-06.bpo-36478.hzyneF.rst]: Would you mind to elaborate the rationale to include "gnu89" C standard and GCC 4.8 version? For example:

"Fix compatibility with ISO C89 needed by "gnu89" standard of GCC 4.8: use C89 for loops in backported pickle patch. Patch by Anthony Sottile."

Modify also the commit message, just reuse the same text (without your name, which is already provided by Git).

Comment (asottile) [reply Misc/NEWS.d/next/Build/2019-03-29-14-29-06.bpo-36478.hzyneF.rst]: done! thanks for the suggestion :)

**Review by vstinner — APPROVED** [review]
LGTM.

I tested manually: compilation fails with gcc -std=gnu89 without the change, but the compilation completes with the change.

**Comment by vstinner** [comment]
@larryhastings: Would you mind to merge this PR to compilation support of old GCC versions?

**Comment by asottile** [comment]
@larryhastings bump on this!

**Comment by larryhastings** [comment]
Thanks for the 3.5 love!

**Pull request merged**